flink-architecture-tests/src/test/java/org/apache/flink/architecture/rules/ApiAnnotationRules.java

```diff
@@ -0,0 +1,168 @@
+
+    /**
+     * This is a stronger requirement than {@link
+     * #PUBLIC_EVOLVING_API_METHODS_USE_ONLY_PUBLIC_EVOLVING_API_TYPES} and ensures that {@link
+     * Public} APIs do not use {@link PublicEvolving} APIs.
+     */
+    @ArchTest
+    public static final ArchRule PUBLIC_API_METHODS_USE_ONLY_PUBLIC_API_TYPES =
+            freeze(
+                    methods()
+                            .that()
+                            .areDeclaredInClassesThat(
+                                    areJavaClasses().and(annotatedWith(Public.class)))
+                            .and()
+                            .arePublic()
+                            .and()
+                            .areNotAnnotatedWith(PublicEvolving.class)
+                            .and()
+                            .areNotAnnotatedWith(Internal.class)
+                            .and()
+                            .areNotAnnotatedWith(Deprecated.class)
+                            .and()
+                            .areNotAnnotatedWith(Experimental.class)
+                            .should(
+                                    haveLeafReturnTypes(
+                                            resideOutsideOfPackage("org.apache.flink..")
+                                                    .or(
+                                                            areDirectlyAnnotatedWithAtLeastOneOf(
+                                                                    Public.class,
+                                                                    Deprecated.class))))
+                            .andShould(
+                                    haveLeafArgumentTypes(
+                                            resideOutsideOfPackage("org.apache.flink..")
+                                                    .or(
+                                                            areDirectlyAnnotatedWithAtLeastOneOf(
+                                                                    Public.class,
+                                                                    Deprecated.class))))
+                            .as(
+                                    "Return and argument types of methods annotated with @Public must be annotated with @Public, too."));
+
+    @ArchTest
+    public static final ArchRule PUBLIC_EVOLVING_API_METHODS_USE_ONLY_PUBLIC_EVOLVING_API_TYPES =
+            freeze(
+                    methods()
+                            .that()
+                            .areNotAnnotatedWith(PublicEvolving.class)
+                            .and()
+                            .areDeclaredInClassesThat(
+                                    areJavaClasses()
+                                            .and(
+                                                    areDirectlyAnnotatedWithAtLeastOneOf(
+                                                            Public.class, PublicEvolving.class)))
```

Comment:
I don't quite recall our previous discussions, but this looks a bit odd. Why are we excluding methods that are annotated as PublicEvolving? Why do we also check classes that are Public (combined these should only cover Public methods)?

If I have a PublicEvolving class, and an explicit PublicEvolving method, then will it not be ignored? Similar case for a Public class and a PublicEvolving method.

I expected something like
`
if annotatedWith(PublicEvolving) || residesInClassWithAnnotation(PublicEvolving.class)
`

Reply:
D'oh, you're absolutely right. Fixed.